browser/ui/views/brave_shields/cookie_list_opt_in_bubble_host.cc

```diff
@@ -0,0 +1,165 @@
+  DCHECK(regional_service_manager);
+  if (regional_service_manager->IsFilterListEnabled(kCookieListUuid)) {
+    return false;
+  }
+
+  return true;
+}
+
+void ShowBubbleOnSessionRestore(base::WeakPtr<Browser> browser, Profile*, int) {
+  if (!browser) {
+    return;
+  }
+
+  auto* bubble_host = CookieListOptInBubbleHost::FromBrowser(browser.get());
+  if (!bubble_host) {
+    return;
+  }
+
+  bubble_host->ShowBubble();
+}
+
+}  // namespace
+
+CookieListOptInBubbleHost::CookieListOptInBubbleHost(Browser* browser)
+    : BrowserUserData<CookieListOptInBubbleHost>(*browser) {
+  GetBrowser().tab_strip_model()->AddObserver(this);
+}
+
+CookieListOptInBubbleHost::~CookieListOptInBubbleHost() = default;
+
+void CookieListOptInBubbleHost::MaybeCreateForBrowser(Browser* browser) {
+  if (ShouldEventuallyShowBubble()) {
+    CreateForBrowser(browser);
+  }
+}
+
+void CookieListOptInBubbleHost::ShowBubble() {
+  // Clear any active session restore callback.
+  session_restored_subscription_ = {};
+
+  if (bubble_manager_ && bubble_manager_->GetBubbleWidget()) {
+    return;
+  }
+
+  if (!ShouldEventuallyShowBubble()) {
+    // If we no longer need to show the opt-in bubble clean up this instance. At
+    // this point, the `this` pointer is no longer valid.
+    RemoveFromBrowser(&GetBrowser());
+    return;
+  }
+
+  // Do not show the bubble if this is not the currently active browser window.
+  auto* browser_view = BrowserView::GetBrowserViewForBrowser(&GetBrowser());
+  if (!browser_view || !browser_view->IsActive()) {
+    return;
+  }
+
+  // Do not show the bubble if the filter list is not yet available, likely
+  // because the regional filter list has not yet been donwloaded.
+  auto* regional_service_manager =
+      g_brave_browser_process->ad_block_service()->regional_service_manager();
+  DCHECK(regional_service_manager);
+  if (!regional_service_manager->IsFilterListAvailable(kCookieListUuid)) {
+    return;
+  }
+
+  // Ensure that the opt-in bubble will not be shown again.
+  g_browser_process->local_state()->SetBoolean(
+      prefs::kAdBlockCookieListOptInShown, true);
+
+  if (!bubble_manager_) {
+    bubble_manager_ = std::make_unique<WebUIBubbleManagerT<CookieListOptInUI>>(
+        GetAnchorView(&GetBrowser()), GetBrowser().profile(),
+        GURL(kCookieListOptInURL), IDS_BRAVE_SHIELDS);
+  }
+
+  if (!bubble_manager_->GetBubbleWidget()) {
+    bubble_manager_->ShowBubble();
+  }
+}
+
+void CookieListOptInBubbleHost::TabChangedAt(content::WebContents* web_contents,
+                                             int index,
+                                             TabChangeType change_type) {
+  // Exit if this is not a "loaded" event for the active tab in this tab strip.
+  if (GetBrowser().tab_strip_model()->active_index() != index ||
+      change_type != TabChangeType::kLoadingOnly || web_contents->IsLoading()) {
+    return;
+  }
+
+  // If a session is in the middle of restoring, attach a callback to show the
```

Comment:
remove tab observer?

Reply:
I've change it so that we exit out if we have an active session restore subscription. This will allow us to handle the edge case where session restore completes but there is no active window. In that case, we'll still be listening for the next "tab load" event.